metricbeat/module/prometheus/query/data.go

```diff
@@ -0,0 +1,161 @@
+)
+
+// ArrayResponse is for "scalar", "string" type.
+type ArrayResponse struct {
+	Status string    `json:"status"`
+	Data   arrayData `json:"data"`
+}
+type arrayData struct {
+	ResultType string        `json:"resultType"`
+	Results    []interface{} `json:"result"`
+}
+
+// MapResponse is for "vector", "matrix" type from Prometheus Query API Request
+type MapResponse struct {
+	Status string  `json:"status"`
+	Data   mapData `json:"data"`
+}
+type mapData struct {
+	ResultType string      `json:"resultType"`
+	Results    []mapResult `json:"result"`
+}
+type mapResult struct {
+	Metric  map[string]string `json:"metric"`
+	Vector  []interface{}     `json:"value"`
+	Vectors [][]interface{}   `json:"values"`
+}
+
+func parseResponse(body []byte, pathConfig QueryConfig) ([]mb.Event, error) {
+	var events []mb.Event
+	var resultType string
+	var convertedMap MapResponse
+
+	// try to convert to array response
+	convertedArray, err := convertJSONToArrayResponse(body)
+	if err != nil {
+		return events, err
+	}
+	resultType = convertedArray.Data.ResultType
+
+	// check if it is a vector or matrix and unmarshal more
+	if resultType == "vector" || resultType == "matrix" {
+		convertedMap, err = convertJSONToMapResponse(body)
+		if err != nil {
+			return events, err
+		}
+		resultType = convertedMap.Data.ResultType
+	}
+
+	switch resultType {
+	case "scalar", "string":
+		if convertedArray.Data.Results != nil {
+			events = append(events, mb.Event{
+				Timestamp: getTimestamp(convertedArray.Data.Results[0].(float64)),
+				MetricSetFields: common.MapStr{
+					"dataType":           resultType,
+					pathConfig.QueryName: attemptConvertToNumeric(convertedArray.Data.Results[1].(string)),
+				},
+			})
+		} else {
+			return events, errors.New("Could not retrieve results")
+		}
+	case "vector":
+		for _, result := range convertedMap.Data.Results {
+			if result.Vector != nil {
+				events = append(events, mb.Event{
+					Timestamp:    getTimestamp(result.Vector[0].(float64)),
+					ModuleFields: common.MapStr{"labels": result.Metric},
+					MetricSetFields: common.MapStr{
+						"dataType":           resultType,
+						pathConfig.QueryName: attemptConvertToNumeric(result.Vector[1].(string)),
+					},
+				})
+			} else {
+				return events, errors.New("Could not retrieve results")
+			}
+		}
+	case "matrix":
+		for _, result := range convertedMap.Data.Results {
+			for _, vector := range result.Vectors {
+				if vector != nil {
+					events = append(events, mb.Event{
+						Timestamp:    getTimestamp(vector[0].(float64)),
+						ModuleFields: common.MapStr{"labels": result.Metric},
+						MetricSetFields: common.MapStr{
+							"dataType":           resultType,
+							pathConfig.QueryName: attemptConvertToNumeric(vector[1].(string)),
+						},
+					})
+				} else {
+					return events, errors.New("Could not retrieve results")
+				}
+			}
+		}
+	default:
+		return events, errors.New("Unknown resultType " + resultType)
```

Comment:
nit: In general, one thing I like in errors is to separate error text from variable content. If for whatever reason `resultType` is empty the error in the console will look like `Unknown resultType` so I usually write variable content in single quotes so the text would become `Unknown resultType ''`. In the second case, while debugging some weird error, you say "What do you mean by ''?". In the first case you think that the error ended there.

It also helps if for whatever reason the variable has a space at the end so you don't end up with things that look like `Unknown resultType matrix` in the console when, in fact, the error was `Unknown resultType 'matrix '`.

This thingy has saved me a lot of headaches